cmd/containerd-shim-runhcs-v1/clone.go

```diff
@@ -0,0 +1,196 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"fmt"
+
+	"github.com/Microsoft/hcsshim/internal/regstate"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+)
+
+const (
+	configRoot = "croot"
+	configKey  = "ckey"
+)
+
+type PersistedUVMConfig struct {
+	// actual information related to template / clone
+	RawData []byte
+	// metadata field used to determine if this config is already started.
+	Stored bool
+}
+
+// loadPersistedConfig loads a persisted config from the registry that matches the given ID
+// If not found returns `regstate.NotFoundError`
+func loadPersistedUVMConfig(ID string) (*PersistedUVMConfig, error) {
+	sk, err := regstate.Open(configRoot, false)
+	if err != nil {
+		return nil, err
+	}
+	defer sk.Close()
+
+	puc := PersistedUVMConfig{
+		Stored: true,
+	}
+	if err := sk.Get(ID, configKey, &puc); err != nil {
+		return nil, err
+	}
+	return &puc, nil
+}
+
+// storePersistedUVMConfig stores or updates the in-memory config to its registry state.
+// If the store fails returns the store error.
+func storePersistedUVMConfig(ID string, puc *PersistedUVMConfig) error {
+	sk, err := regstate.Open(configRoot, false)
+	if err != nil {
+		return err
+	}
+	defer sk.Close()
+
+	if puc.Stored {
+		if err := sk.Set(ID, configKey, puc); err != nil {
+			return err
+		}
+	} else {
+		if err := sk.Create(ID, configKey, puc); err != nil {
+			return err
+		}
+	}
+	puc.Stored = true
+	return nil
+}
+
+// Remove removes any persisted state associated with this config. If the config
+// is not found in the registery `Remove` returns no error.
+func removePersistedUVMConfig(ID string) error {
+	sk, err := regstate.Open(configRoot, false)
+	if err != nil {
+		if regstate.IsNotFoundError(err) {
+			return nil
+		}
+		return err
+	}
+	defer sk.Close()
+
+	if err := sk.Remove(ID); err != nil {
+		if regstate.IsNotFoundError(err) {
+			return nil
+		}
+		return err
+	}
+	return nil
+}
+
+// When encoding interfaces gob requires us to register the struct types that we will be using under those
+// interfaces. This registration needs to happen on both sides i.e the side which encodes the data and the
+// side which decodes the data.
+func gobInit() {
```

Comment:
More of a nitpick but we could just put this in an init function that will automatically get run. https://golang.org/doc/effective_go.html#init

Reply:
Thanks. This was a nice suggestion.